x-pack/plugins/security_solution/common/detection_engine/schemas/request/add_prepackged_rules_schema.test.ts

```diff
@@ -776,7 +778,9 @@ describe('add prepackaged rules schema', () => {
     const decoded = addPrepackagedRulesSchema.decode(payload);
     const checked = exactCheck(payload, decoded);
     const message = pipe(checked, foldLeftRight);
-    expect(getPaths(left(message.errors))).toEqual(['Invalid value "-1" supplied to ""']);
+    expect(getPaths(left(message.errors))).toEqual([
+      'Invalid value "-1" supplied to "PositiveIntegerGreaterThanZero"',
```

Comment:
I kind of expected it to say which property was wrong, but I also like knowing what the correct type is. Could we add both?

Reply:
Yeah, these are the parts of the error formatter where io-ts doesn't give it the keys which makes it tricky. It gives you the property name though. You can see in most of the validation errors it gives back the correct key such as "note" or "max_signals" etc..

However, I think there is a name and value where I might be able to extract it. It was tricky to get it to give me this information it did. No promises on it, but I will look into it before I merge to see if it's possible to do get the associated key from the value.

I absolutely do agree it would make things clearer to the end user and be a better message if I can pull it off here... Will look into it for a bit.

Reply:
The problem is when I look at the context of the custom types sometimes I don't have enough to do anything within the error reporter. Using this one as an example this is my context when I get an error within `formatErrors`:

```ts
   errors: [
      {
        "value": -1,
        "context": [
          {
            "key": "",
            "type": {
              "name": "PositiveIntegerGreaterThanZero"
            },
            "actual": -1
          }
        ]
      }
    ]
```

I don't know if I am doing something incorrect still with the API at this point when I create the custom types or this is the way it is supposed to be within io-ts where sometimes it does not give you the full context when you write your own custom types.

I still think I'm missing something simple somewhere when I create `PositiveIntegerGreaterThanZero` that does not give it the full context and that's why the error reporting is not giving better output in these cases. I just don't know what yet.

Reply:
I'm not sure if you already found this one but it seems like it addressed what we want, though I'm still not 100% on implementation. https://github.com/gcanti/io-ts/issues/194

Reply:
I think maybe, that third `actual` property is one we supply?

```
decode(i: I): Validation<A> {
    return this.validate(i, [{ key: '', type: this, actual: i }])
}
```

Reply:
Playing around with `PathReporter` that seems to give more context too. For the validator test, instead of returning `'Invalid value "some other value" supplied to "a"'`, it returned `'Invalid value "some other value" supplied to : {| a: number |}/a: number'` 🤷‍♀️

Reply:
Ok, after digging a lot more I think I figured out how to get the context for this and everywhere else and clear these up.

Within our custom types we have to do this:

```ts
  (input, context): Either<t.Errors, number> => {
    return input == null ? t.success(DEFAULT_MAX_SIGNALS) : max_signals.validate(input, context);
  },
```

Rather than this:

```ts
  (input): Either<t.Errors, number> => {
    return input == null ? t.success(DEFAULT_MAX_SIGNALS) : max_signals.decode(input);
  },
```

Doing that validate and passing down the `context` from the top level looks like that is what gets us the keys in the right spots where we need them. I am going to push up the changes to all the types here and within lists now with this fix.

Reply:
Nice! That is awesome. Super helpful and way better than `[object object]` or the empty string we get now 🎉

Reply:
Just pushed it up! You should see the changes now with the latest commit.

Reply:
Not suggesting further changes, but thought this was super helpful - https://dev.to/gnomff_65/fp-ts-and-beautiful-api-calls-1f55